File: Practice/binary_search/binary_search.py
#-------------------------------------------------------------------------------------
# 二分探索
# 引数: x  返り値: 「x以上の中で最も小さい数の位置」
#-------------------------------------------------------------------------------------
n = 5
a = [1, 3, 4, 6, 9]
def Nibutan(x: int) -> int:
    # 左端
    l = 0
    # 右端
    r = n - 1
    
    # l < 右端 - 左端の間
    while l < r - 1:
        # 中央
        c = (l + r) // 2
        # a[c] < x ならば（条件を満たさない場合）
        if a[c] < x:
            l = c
        # そうでなければ
        else:
            # 右端を中央へ更新
            r = c
    # 右端を返す
    return r

# 5を初めて越えるのはあindex(3)
print(Nibutan(5))

File: src/practice/binary_search/binary_search.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    二分探索
    - 目的の値の最小のIndexを返す（ない場合は-1）
    - leftは「常に」条件を満たさない
    - rightは「常に」条件を満たす
*/
bool isOk(vector<int> a, int index, int key) {
    if (a[index] >= key)
        return true;
    else
        return false;
}

int binary_search(vector<int> a, int key) {
    int left = -1;
    int right = (int)a.size();
    while (right - left > 1) {
        int mid = left + (right - left) / 2;
        if (isOk(a, mid, key))
            right = mid;
        else
            left = mid;
    }
    return right;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<int> a = {10, 30, 35, 50, 50, 75};
    int place = binary_search(a, 50);

    cout << place << endl;

    return 0;
}
